internal/engine/dcwatch/disabler.go

```diff
@@ -0,0 +1,126 @@
+package dcwatch
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	"github.com/tilt-dev/tilt/internal/dockercompose"
+	"github.com/tilt-dev/tilt/pkg/logger"
+	"github.com/tilt-dev/tilt/pkg/model"
+)
+
+const disableDebounceDelay = 200 * time.Millisecond
+
+// disables DC resources w/ debouncing to group them into one docker-compose invocation
+type disabler struct {
+	dcc      dockercompose.DockerComposeClient
+	clock    clockwork.Clock
+	updates  chan disableQueueEntry
+	loopOnce sync.Once
+}
+
+func newDisabler(dcc dockercompose.DockerComposeClient, clock clockwork.Clock) disabler {
+	return disabler{
+		dcc:     dcc,
+		clock:   clock,
+		updates: make(chan disableQueueEntry),
+	}
+}
+
+type disableQueueEntry struct {
+	Spec         model.DockerComposeUpSpec
+	NeedsCleanup bool
+	// the container's start time
+	StartTime time.Time
+	// the service's order wrt other services, to ensure logging in consistent order
+	Order int
+}
+
+func (d *disabler) Update(ctx context.Context, e disableQueueEntry) {
+	d.loopOnce.Do(func() {
+		go d.Loop(ctx)
+	})
+
+	d.updates <- e
```

Comment:
if 'docker rm' really takes 10s, aren't we worried about race conditions between when an 'rm' enters the queue and when it gets scheduled? e.g., it seems like if a user is toggling a service on and off, bizarre things could happen. where the 'rm' could get schedules 10s later.

Reply:
> aren't we worried about race conditions between when an 'rm' enters the queue and when it gets scheduled?

yes! IIUC, that's handled correctly, but maybe I missed something? The loop maintains the current state of which resources are enabled and disabled, and when it calls `rm`, only passes resources that are currently disabled (i.e., if a resource is reenabled during those 10 seconds, a new update will come in with `NeedsCleanup: false`, and get filtered out before `rm` is called). Is that the concern you describe?

(Obviously, it's possible that the user reenables the resources after the last state update but before we've invoked `rm`, but that section of code should be ~instant, and that's effectively the same case as enabling the resource while it's in the process of disabling--the behavior in that case is reasonable - it finishes disabling, then reenables)

Reply:
hmmm...maybe? i dunno the scheduling seems sus to me here.

if you use a blocking channel here, then will we have any race conditions where we're holding onto the EngineState Lock while the channel blocks? which means tilt will lock up?

it's also possible that i'm just more used to the pattern that kubernetes uses, where they try to keep the state caching completely separate from the scheduling logic, though i appreciate that approach has pros/cons too.